distributed/worker.py

```diff
@@ -2999,6 +3012,8 @@ async def gather_dep(
                 for d in has_what:
                     ts = self.tasks[d]
                     ts.who_has.remove(worker)
+                    if not ts.who_has:
+                        recommendations[ts] = "missing"
```

Comment:
Logging here might be helpful for future debugging. Probably shouldn't call it `missing-dep`, so it can be differentiated from the `finally` case, but something in the same spirit.

It also might be worth a comment on why we don't send `missing-data` to the scheduler in this case, but do send it in the other case of a missing dep. (Because in this case, we don't know whether the dep is actually missing from the worker—we're just preemptively assuming it is because we're assuming the whole worker has died, but we don't want to send the scheduler wrong information based on that assumption.)

Reply:
I've added logging.  I haven't added the comment.  I felt that I didn't understand the reasoning here sufficiently well to argue for one way or the other.

Reply:
There are two things were this code path becomes relevant

1. Worker is actually dead. The way we implement this handler here acts as if the worker was dead. We basically purge the entire info we know about this worker because we don't expect him to come back. 

The scheduler will detect the dead worker eventually and reschedule the task. By not sending the `missing-data` signal we open ourselves to a minor but harmless race condition where the scheduler distributes "faulty" who_has to other workers for a brief time such that multiple workers may run into this OSError. That's unfortunate from a performance perspective, particularly if they configure very high connect timeouts. The "correct" way would be to send the signal but it shouldn't matter a lot in the end.

2. There was a network blip. We should actually not do anything in case of a network blip but just retry (w/ backoff). Sending a missing-data in this case might actually be very harmful since the scheduler removes this worker then from its who_has and the worker will never receive the "free-keys" signal, i.e. we'd acquire zombie tasks.

We currently cannot distinguish 1. and 2. so we need to find a middle ground. Purging data is safe because we can reacquire this information. Sending `missing-data` in the wrong situation has the potential for being unsafe so we should not do it and live with this very minor race condition mentioned above

Reply:
> We currently cannot distinguish 1. and 2. so we need to find a middle ground. Purging data is safe because we can reacquire this information. Sending missing-data in the wrong situation has the potential for being unsafe so we should not do it and live with this very minor race condition mentioned above

This explanation is basically what I was looking for in the comment.